### PR TITLE
Refactor conversations

### PR DESCRIPTION
**Pull request by tiltec**
- rewrite conversations module in fcm-style and withMeta
- split Conversations components from Wall
- rename WallInput to ConversationCompose
- rename WallMessage to ConversationMessage

Should work, although I didn't do any testing of the error handling.

### CODE REVIEW

**Comment by tiltec** [comment]
This `CMessage` looks weird to me now, maybe I rename to `ConversationMessage`?

**Comment by nicksellen** [comment]
> This CMessage looks weird to me now, maybe I rename to ConversationMessage?

Yes, and fits the style guide point that you should name components that are only relevant within another one as a prefix e.g. `Conversation` and `ConversationMessage`

File: src/store/modules/conversations.js [hunk 1]
```diff
@@ -1,243 +1,119 @@
...
+      return {
+        messages: getters.activeMessages,
+        canLoadMore,
+        ...metaStatusesWithId(getters, ['send', 'fetch', 'fetchMore'], id),
+      }
+    },
   },
+  actions: {
+    ...withMeta({
+      async send ({ commit, state, dispatch }, conversationId, messageData) {
```
Comment (nicksellen) [inline src/store/modules/conversations.js]: Oh, I didn't know you pass two args? Is it supported?

File: src/store/modules/conversations.js [hunk 2]
```diff
@@ -1,243 +1,119 @@
...
-    }
-  },
-  [types.RECEIVE_MESSAGES_ERROR] (state, { conversationId, error }) {
-    console.log(error)
-    if (state.activeConversationId === conversationId) {
-      state.receiveStatus = {
-        isWaiting: false,
-        error,
-        success: false,
+      if (!getters.activeMessages.find(e => e.id === message.id)) {
```
Comment (nicksellen) [inline src/store/modules/conversations.js]: A bit inefficient to go through all the messages each time, ok for now I'm sure but could be much improved.

File: src/store/modules/conversations.js [hunk 3]
```diff
@@ -1,243 +1,119 @@
...
+    setMessages (state, { conversationId, messages, cursor }) {
+      Vue.set(state.messages, conversationId, messages)
       Vue.set(state.cursors, conversationId, cursor)
-    }
-    if (state.activeConversationId === conversationId) {
-      state.receiveMoreStatus = {
-        isWaiting: false,
-        error: null,
+    },
+    appendMessages (state, { conversationId, messages, cursor }) {
```
Comment (nicksellen) [inline src/store/modules/conversations.js]: Not sure I quite understand why we would prepend and append messages.

File: src/components/Conversation/CMessage.vue [hunk 4]
```diff
@@ -25,7 +25,7 @@ import { QItem, QItemSide, QItemMain, QItemTile } from 'quasar'
 import DateAsWords from '@/components/General/DateAsWords'
 
 export default {
-  name: 'WallMessage',
+  name: 'CMessage',
```
Comment (nicksellen) [inline src/components/Conversation/CMessage.vue]: `ConversationMessage` fits the styleguide rule better.

File: src/components/Conversation/Compose.vue [hunk 5]
```diff
@@ -22,12 +22,13 @@
 <script>
 import ProfilePicture from '@/components/ProfilePictures/ProfilePicture'
 import { QItem, QItemMain, QInput, QBtn, QItemSide, QItemTile } from 'quasar'
+import statusMixin from '@/mixins/statusMixin'
 
 export default {
-  name: 'WallInput',
+  name: 'Compose',
```
Comment (nicksellen) [inline src/components/Conversation/Compose.vue]: `ConversationCompose` might fit better, or could this also be a more standalone component?

File: src/components/Conversation/CMessage.vue (hunk 4 above)
Comment (tiltec) [reply src/components/Conversation/CMessage.vue]: done

File: src/components/Conversation/Compose.vue (hunk 5 above)
Comment (tiltec) [reply src/components/Conversation/Compose.vue]: done

File: src/store/modules/conversations.js (hunk 1 above)
Comment (tiltec) [reply src/store/modules/conversations.js]: I thought it works, then I tried again, it fails. Will pass an object instead.

File: src/store/modules/conversations.js (hunk 2 above)
Comment (tiltec) [reply src/store/modules/conversations.js]: I agree, we could map the messages by ID, store in an object and then do a simple key lookup. But I'm sure that this is not the reason why the page is sluggish sometimes ;)

File: src/store/modules/conversations.js (hunk 3 above)
Comment (tiltec) [reply src/store/modules/conversations.js]: prepend: when the user writes a message and it should go to top
append: when loading more message history from the server

will add comments :)

File: src/store/modules/conversations.js (hunk 2 above)
Comment (nicksellen) [reply src/store/modules/conversations.js]: I was thinking about an  "always sorted list" approach (would also fit with de-duplication)

**Pull request merged**